### ISSUE TITLE
Jumps in stack limits?

### ISSUE THREAD

**Issue by sol** [ndmitchell/spaceleak#3]
The README uses `-K33K` as an example for a stack limit that is required for a program to run.  And while I experimented with this technique I observed the exact same limit a couple of times.

I don't think this is a coincidence!

More specifically, I repeatedly had situations where a program completed successfully with `-K1K`.  When I then incrementally increased the input size it would continue to work until at some point I would observe a jump where it would require `-K33K`.

One explanation for this would be that stack limits are enforced in `32K` blocks.

An other thought, I haven't used `join . onceFork`.  @ndmitchell in the README you mention a minimum stack limit of `8K` for the main thread.  Do you know if this was for `x86`?  Could it be that for `x86_64` this is actually `32K`?

**Comment by ndmitchell** [comment]
You're definitely on to something - I think you can change the increments of stack pieces perhaps? Although I think there are various minimums. It definitely doesn't work linearly.

**Comment by ndmitchell** [comment]
I added `measureStack` to the file SpaceLeak.hs in this repo, and with that I can actually measure the effective stack. It's complicated, but if you add `-kc2k` then the stack increases in a much smaller granularity and it becomes easier to understand. I'm slightly tempted to add that to the instructions, but will play with Shake first.

**Comment by ndmitchell** [comment]
I tried this technique with Shake, found one small place that strictness made sense and improved stack usage at https://github.com/ndmitchell/shake/commit/0f41895bebea0e353490750ff402df277b3810ee, and found a GHC bug: https://ghc.haskell.org/trac/ghc/ticket/15595

**Comment by sol** [comment]
Regarding that GHC bug, I'm sure I have observed something similar before.  Instead of getting a stack overflow the program would hog one core and start allocating more and more memory.  Can't remember how to reproduce.

**Comment by sol** [comment]
I suspect this call to `bracket_` here: https://hackage.haskell.org/package/base-4.11.1.0/docs/src/System.Environment.html#withProgArgv

Edit: Actually there multiple calls to `bracket` involved via `withCString`.

**Comment by ndmitchell** [comment]
Yep, it's all documented on the GHC Trac now, along with a smaller test case.

**Comment by ndmitchell** [comment]
Documented in the ticket. I tried applying this to Shake, and caught a few nice improvements, but got stuck when I discovered that break in profiling mode at `-O1` (which the base libraries are compiled with) has a space leak I can't understand. That makes the `-xc` trick useless, since it requires profiling, and it ground to a halt. I suspect `-K1K` is small enough to catch anything serious, and anything more is not so useful.
